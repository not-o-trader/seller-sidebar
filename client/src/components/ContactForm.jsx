import React from 'react';
import styled from 'styled-components';
import { formatMoney } from 'accounting';

import FormInput from './FormInput.jsx';

const Wrapper = styled.div`
  background-color: #e6e6e6;
  border: 1px solid #d4d4d4;
  margin-bottom: 15px;
`;

const Inner = styled.div`
  width: 95%;
  margin: 0 auto;
`;

const Label = styled.label`
  display: block;
  margin: 10px 0;
  font-weight: bold;
`;

const TextArea = styled.textarea`
  width: 100%;
  display: block;
  font-size: 14px;
  line-height: 21px;
`;

const Float = styled.div`
  float: left;
`;

const Button = styled.button`
  width: 100%;
  padding: 7px 14px;
  font-size: 14px;
  display: block;
  color: white;
  background-color: rgb(232, 118, 0);
  line-height: 21px;
  box-shadow: rgb(0, 0, 0, 0.1);
  border-bottom-left-radius: 3.5px;
  border-bottom-right-radius: 3.5px;
  border-top-left-radius: 3.5px;
  border-top-right-radius: 3.5px;
`;

const Input = styled.input`
  margin-right: 20px;
`;

class ContactForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      message: '',
      firstName: '',
      lastName: '',
      email: '',
      phone: '',
      enroll: false
    };

    this.handleMessageChange = this.handleChange.bind(this, 'message');
    this.handleFirstNameChange = this.handleChange.bind(this, 'firstName');
    this.handleLastNameChange = this.handleChange.bind(this, 'lastName');
    this.handleEmailChange = this.handleChange.bind(this, 'email');
    this.handlePhoneChange = this.handleChange.bind(this, 'phone');
    this.handleEnrollChange = this.handleChange.bind(this, 'enroll');
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  static getDerivedStateFromProps(props, state) {
    if (state && state.message === '') {
      const { year, make, model, price } = props;
      if (year && make && model && price) {
        const priceStr = formatMoney(price);
        const message = `
        Hi, I'm interested in your ${year} ${make} ${model} listed on Not-O-Trader
        for ${priceStr}. I'd like to get more information about this vehicle and
        confirm its availability.
        `.trim().replace(/[\n ]+/g, ' ');
        return { ...state, message };
      }
    }
    return state;
  }

  handleChange(prop, event) {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    this.setState({ [prop]: value });
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.send(this.state);
    this.setState({
      message: '',
      firstName: '',
      lastName: '',
      email: '',
      phone: '',
      enroll: false
    });
  }

  render() {
    const placeholder = 'Enter your message here. Maximum 250 characters';
    return (
      <Wrapper>
        <Inner>
          <h2>Email Seller</h2>
          <form onSubmit={this.handleSubmit} autoComplete="on">
            <Label htmlFor="message">Message</Label>
            <TextArea
              name="message"
              maxLength={250}
              rows={8}
              placeholder={placeholder}
              value={this.state.message}
              onChange={this.handleMessageChange}
            >
            </TextArea>
            <FormInput
              name="firstName"
              label="First Name"
              type="text"
              placeholder="First Name"
              value={this.state.firstName}
              autocomplete="given-name"
              handleChange={this.handleFirstNameChange}
            />
            <FormInput
              name="lastName"
              label="Last Name"
              type="text"
              placeholder="Last Name"
              value={this.state.lastName}
              autocomplete="family-name"
              handleChange={this.handleLastNameChange}
            />
            <FormInput
              name="emailAddress"
              label="Email"
              type="email"
              placeholder="Email"
              value={this.state.email}
              autocomplete="email"
              handleChange={this.handleEmailChange}
            />
            <FormInput
              name="phoneNumber"
              label="Phone Number"
              type="tel"
              placeholder=""
              optional="true"
              value={this.state.phone}
              autocomplete="tel"
              handleChange={this.handlePhoneChange}
            />
            <div>
              <Float>
                <Input
                  type="checkbox"
                  name="contactCheckbox"
                  value={this.state.enroll}
                  onChange={this.handleEnrollChange}
                >
                </Input>
              </Float>
              <div>
                Yes, I would like to receive updates, promotions, and discounts
                from Not-O-Trader and their affiliates.
              </div>
            </div>
            <div>
              <p>
                By using this service, you accept the terms of
                our Visitor Agreement
              </p>
            </div>
            <Button
              type="submit"
              id="send-seller-email"
            >
              Submit
            </Button>
          </form>
        </Inner>
      </Wrapper>
    );
  }
}

export default ContactForm;
