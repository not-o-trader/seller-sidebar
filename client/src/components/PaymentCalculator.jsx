import React from 'react';
import styled from 'styled-components';
import { formatMoney } from 'accounting';

import FormInput from './FormInput.jsx';

const Wrapper = styled.div`
  position: absolute;
  width: 95%;
  top: 10px;
  left: 50%;
  transform: translateX(-50%);
  border: 1px solid #d4d4d4;
  background-color: white;
`;

const Inner = styled.div`
  width: 95%;
  margin: 10px auto;
`;

const Header = styled.div`
  margin-top: 10px;
`;

const Label = styled.div`
  margin: 10px 0;
  font-weight: bold;
`;

const Calculated = styled.p`
  text-align: right;
`;

const CloseButton = styled.button`
  float: right;
`;

const MonthsButton = styled.button`
  color: ${props => props.selected ? 'white' : 'blue'};
  background-color: ${props => props.selected ? 'blue' : 'white'};
  outline-width: 0;
`;

const ResetButton = styled.button`
  padding: 7px 14px;
  margin-left: 40%;
  font-size: 14px;
  color: blue;
  line-height: 21px;
  box-shadow: rgb(0, 0, 0, 0.1);
  border-bottom-left-radius: 3.5px;
  border-bottom-right-radius: 3.5px;
  border-top-left-radius: 3.5px;
  border-top-right-radius: 3.5px;
`;

const CalculateButton = styled.button`
  padding: 7px 14px;
  margin: auto 10px;
  font-size: 14px;
  color: white;
  background-color: rgb(232, 118, 0);
  line-height: 21px;
  box-shadow: rgb(0, 0, 0, 0.1);
  border-bottom-left-radius: 3.5px;
  border-bottom-right-radius: 3.5px;
  border-top-left-radius: 3.5px;
  border-top-right-radius: 3.5px;
`;

class PaymentCalculator extends React.Component {
  constructor(props) {
    super(props);
    this.initialState = {
      months: 60,
      interestRate: 0,
      price: props.price || 0,
      downPayment: 0,
      tradeInValue: 0,
      amountOwedOnTrade: 0,
      estimatedPayment: 0,
      totalFinanced: 0,
      totalInterest: 0,
      totalLoan: 0
    };
    this.state = { ...this.initialState };
    this.handleClose = this.handleClose.bind(this);
    this.handleReset = this.handleReset.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleRateChange = this.handleChange.bind(this, 'interestRate');
    this.handlePriceChange = this.handleChange.bind(this, 'price');
    this.handleDownPaymentChange = this.handleChange.bind(this, 'downPayment');
    this.handleTradeChange = this.handleChange.bind(this, 'tradeInValue');
    this.handleOwedChange = this.handleChange.bind(this, 'amountOwedOnTrade');
    this.handleMonthsClick = this.handleMonthsClick.bind(this);
  }

  static getDerivedStateFromProps(props, state) {
    const price = props.price || state.price;
    if (typeof price === 'number') {
      // const downPayment = price * 0.2;
      return { ...state, price };
    }
    return state;
  }

  recalculate() {
    const { price, downPayment, tradeInValue, amountOwedOnTrade } = this.state;
    const loan = price - downPayment - tradeInValue + amountOwedOnTrade;
    const payment = Math.ceil(loan / this.state.months);
    // TODO: Interest etc.
    this.setState({
      totalFinanced: Math.ceil(loan),
      totalLoan: Math.ceil(loan), // FIXME
      estimatedPayment: Math.ceil(payment)
    });
  }

  handleClose(e) {
    e.preventDefault();
    e.stopPropagation();
    this.props.toggle();
  }

  handleReset(e) {
    e.preventDefault();
    e.stopPropagation();
    const state = {
      ...this.initialState,
      price: this.props.price || 0,
      totalFinanced: 0,
      totalLoan: 0,
      estimatedPayment: 0
    };
    this.setState(state, () => this.recalculate());
  }

  handleSubmit(e) {
    e.preventDefault();
    e.stopPropagation();
    this.recalculate();
  }

  handleChange(prop, event) {
    const target = event.target;
    let value = target.type === 'checkbox' ? target.checked : target.value;
    if (typeof value === 'string') {
      if (value.match(/^[0-9]*$/)) {
        value = +value;
      } else {
        console.warn(`Non-numeric value entered for ${prop}`);
      }
    }
    this.setState({ [prop]: value }, () => this.recalculate());
  }

  handleMonthsClick(event) {
    event.preventDefault();
    event.stopPropagation();
    const value = event.target.value;
    this.setState({ months: +value });
  }

  render() {
    return (
      <Wrapper id="payment-calculator-modal">
        <Inner>
          <div>
            <Header>
              Calculator
              <CloseButton onClick={this.handleClose}>X</CloseButton>
            </Header>
          </div>
          <hr />
          <form>
            <div>
              <Label>Months</Label>
              <div>
                <MonthsButton
                  value="24"
                  onClick={this.handleMonthsClick}
                  selected={this.state.months === 24}
                >
                  24
                </MonthsButton>
                <MonthsButton
                  value="36"
                  onClick={this.handleMonthsClick}
                  selected={this.state.months === 36}
                >
                  36
                </MonthsButton>
                <MonthsButton
                  value="48"
                  onClick={this.handleMonthsClick}
                  selected={this.state.months === 48}
                >
                  48
                </MonthsButton>
                <MonthsButton
                  value="60"
                  onClick={this.handleMonthsClick}
                  selected={this.state.months === 60}
                >
                  60
                </MonthsButton>
                <MonthsButton
                  value="72"
                  onClick={this.handleMonthsClick}
                  selected={this.state.months === 72}
                >
                  72
                </MonthsButton>
              </div>
            </div>
            <FormInput
              name="interestRate"
              label="Interest Rate"
              type="number"
              value={this.state.interestRate}
              handleChange={this.handleRateChange}
            />
            <FormInput
              name="price"
              label="Price"
              type="number"
              value={this.state.price}
              handleChange={this.handlePriceChange}
            />
            <FormInput
              name="downPayment"
              label="Down Payment"
              type="number"
              value={this.state.downPayment}
              handleChange={this.handleDownPaymentChange}
            />
            <FormInput
              name="tradeInValue"
              label="Trade In Value"
              type="number"
              value={this.state.tradeInValue}
              handleChange={this.handleTradeChange}
            />
            <p>
              <a href='#'>Get your Kelly Blue Book Instant Cash Offer today</a>
            </p>
            <FormInput
              name="amountOwedOnTrade"
              label="Amount Owed on Trade"
              type="number"
              value={this.state.amountOwedOnTrade}
              handleChange={this.handleOwedChange}
            />
            <div>
              <p>Estimated Payment*</p>
              <p>
                ${this.state.estimatedPayment} / month for {this.state.months} at {this.state.interestRate}% APR.
              </p>
            </div>
            <div>
              <p>Price</p>
              <Calculated>{formatMoney(this.state.price)}</Calculated>
            </div>
            <div>
              <p>Down Payment</p>
              <Calculated>- {formatMoney(this.state.downPayment)}</Calculated>
            </div>
            <div>
              <p>Trade-In Value</p>
              <Calculated>- {formatMoney(this.state.tradeInValue)}</Calculated>
            </div>
            <div>
              <p>Amount Owed on Trade</p>
              <Calculated>{formatMoney(this.state.amountOwedOnTrade)}</Calculated>
            </div>
            <div>
              <p>Total Financed</p>
              <Calculated>{formatMoney(this.state.totalFinanced)}</Calculated>
            </div>
            <div>
              <p>Total Interest</p>
              <Calculated>{formatMoney(this.state.totalInterest)}</Calculated>
            </div>
            <div>
              <p>Total Loan</p>
              <Calculated>{formatMoney(this.state.totalLoan)}</Calculated>
            </div>
            <p>
              *Estimated payments are for informational purposes only and don't
              account for tax, title and fees or represent a guarantee of credit
              from the seller. Such estimates are not an offer or commitment to
              lend.
            </p>
            <div>
              <ResetButton
                id="payment-calculator-reset"
                onClick={this.handleReset}
              >
                Reset
              </ResetButton>
              <CalculateButton
                id="payment-calculator-submit"
                onClick={this.handleSubmit}
              >
                Calculate
              </CalculateButton>
            </div>
          </form>
        </Inner>
      </Wrapper>
    );
  }
}

export default PaymentCalculator;
