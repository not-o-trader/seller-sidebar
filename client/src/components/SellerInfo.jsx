import React from 'react';
import styled from 'styled-components';

const Wrapper = styled.div`
  background-color: #e6e6e6;
  border: 1px solid #d4d4d4;
  margin-bottom: 15px;
`;

const Inner = styled.div`
  width: 95%;
  margin: 0 auto;
`;

const Phone = styled.p`
  font-weight: bold;
`;

const Float = styled.div`
  float: left;
  margin-right: 10px;
`;

const Button = styled.button`
  width: 100%;
  padding: 7px 14px;
  font-size: 14px;
  display: block;
  line-height: 21px;
  box-shadow: rgb(0, 0, 0, 0.1);
  border-bottom-left-radius: 3.5px;
  border-bottom-right-radius: 3.5px;
  border-top-left-radius: 3.5px;
  border-top-right-radius: 3.5px;
`;

const SellerInfo = ({ logo, name, phone, address, city, state, zip, website }) => {
  const mapsAddr = `${address}, ${city}, ${state} ${zip}`;
  const mapsLink = `https://maps.google.com/maps/place/${mapsAddr}`;
  return (
    <Wrapper>
      <Inner>
        {logo && <img src={logo} alt={name} />}
        <h2>{name}</h2>
        <div>
          <Phone>{phone}</Phone>
        </div>
        <div>
          <Float>
            <a href={mapsLink}>
              <span className="fas fa-map-marker-alt"></span>
            </a>
          </Float>
          <div>
            <p>
              <a href={mapsLink}>
                {address} {city}, {state} {zip}
              </a>
            </p>
          </div>
        </div>
        <div>
          <p>
            <a href="#">View our inventory</a>
          </p>
        </div>
        <div>
          <p>
            <Button>
              <a href={website}>Visit Dealer Website <span className="fas fa-external-link-alt"></span></a>
            </Button>
          </p>
        </div>
      </Inner>
    </Wrapper>
  );
};

export default SellerInfo;
