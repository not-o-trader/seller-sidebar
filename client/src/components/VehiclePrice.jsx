import React from 'react';
import styled from 'styled-components';
import { formatMoney } from 'accounting';

const Wrapper = styled.div`
  width: 95%;
  margin: 0 auto;
`;

const FinancingLink = styled.a`
  margin-right: 10px;
`;

const PaymentsLink = styled.a`
  float: right;
`;

const KBB = styled.div`
  margin-top: -8px;
  float: left;
`;

const FMR = styled.div`
  margin: auto 0;
`;

class VehiclePrice extends React.Component {
  constructor(props) {
    super(props);
    this.handleCalculatorClick = this.handleCalculatorClick.bind(this);
  }

  handleCalculatorClick(e) {
    e.preventDefault();
    this.props.toggleModal();
  }

  render() {
    return (
      <Wrapper>
        <h3>{formatMoney(this.props.price)}</h3>
        <div>
          <div>
            <KBB>
              <img src="KBB_Badge.png" />
            </KBB>
            <FMR>
              <a href='#'>Is this price in the Fair Market Range?</a>
            </FMR>
          </div>
          <div>
            <p>
              <FinancingLink href='#'>$ Arrange Financing <span className="fas fa-external-link-alt"></span></FinancingLink>
              <PaymentsLink
                href='#'
                id="show-payment-calculator"
                onClick={this.handleCalculatorClick}
              >
                <span className="fas fa-calculator"></span> Calculate payment
              </PaymentsLink>
            </p>
          </div>
        </div>
      </Wrapper>
    );
  }
}

export default VehiclePrice;
