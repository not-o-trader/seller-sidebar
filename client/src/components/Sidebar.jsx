import React from 'react';
import axios from 'axios';
import styled from 'styled-components';

import VehiclePrice from './VehiclePrice.jsx';
import SellerInfo from './SellerInfo.jsx';
import ContactForm from './ContactForm.jsx';

const Wrapper = styled.div`
  max-width: 100%;
  font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif;
  font-size: 0.875em;
  a:link, a:visited, a:hover, a:active {
    color: blue;
  }
`;

class Sidebar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
    this.sendSellerEmail = this.sendSellerEmail.bind(this);
  }

  componentDidMount() {
    this.getInfo(1);
  }

  getInfo(vehicleId) {
    axios.get(`/api/seller/${vehicleId}`)
      .then(({ data }) => {
        this.setState(data);
      })
      .catch((error) => {
        console.log('Error getting data:', error);
      });
  }

  sendSellerEmail(data) {
    const extra = {
      vehicleId: this.state.vehicleId,
      sellerEmail: this.state.sellerEmail
    };
    data = Object.assign({}, data, extra);
    axios.post('/api/seller/email', data)
      .then(res => console.log('Successfully sent email'))
      .catch(err => console.error('Error sending email:', err));
  }

  render() {
    return (
      <Wrapper>
        <VehiclePrice price={this.state.vehiclePrice} />
        <hr />
        <SellerInfo
          logo={this.state.sellerLogoUrl}
          name={this.state.sellerName}
          phone={this.state.sellerPhone}
          address={this.state.sellerAddress}
          city={this.state.sellerCity}
          state={this.state.sellerState}
          zip={this.state.sellerZip}
          website={this.state.sellerWebsite}
        />
        <ContactForm
          make={this.state.vehicleMake}
          model={this.state.vehicleModel}
          year={this.state.vehicleModelYear}
          email={this.state.sellerEmail}
          price={this.state.vehiclePrice}
          send={this.sendSellerEmail}
        />
      </Wrapper>
    );
  }
}

export default Sidebar;
