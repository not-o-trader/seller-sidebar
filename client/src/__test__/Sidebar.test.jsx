import React from 'react';
import { shallow, mount } from 'enzyme';
import axios from 'axios';

import Sidebar from '../components/Sidebar.jsx';
import VehiclePrice from '../components/VehiclePrice';
import SellerInfo from '../components/SellerInfo.jsx';
import ContactForm from '../components/ContactForm.jsx';
import PaymentCalculator from '../components/PaymentCalculator';

jest.mock('axios');

describe('<Sidebar />', () => {
  axios.get.mockResolvedValue({});
  axios.post.mockResolvedValue({});
  const shallowWrapper = shallow(<Sidebar />);
  const mountWrapper = mount(<Sidebar />);

  it('renders a <VehiclePrice /> component', () => {
    expect(shallowWrapper.find(VehiclePrice).length).toBe(1);
  });

  it('renders a <SellerInfo /> component', () => {
    expect(shallowWrapper.find(SellerInfo).length).toBe(1);
  });

  it('renders a <ContactForm /> component', () => {
    expect(shallowWrapper.find(ContactForm).length).toBe(1);
  });

  it('renders', () => {
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('does not render a <PaymentCalculator /> component', () => {
    expect(shallowWrapper.find(PaymentCalculator).length).toBe(0);
  });

  it('renders a <PaymentCalculator /> component on demand', () => {
    mountWrapper.find('a#show-payment-calculator').simulate('click');
    expect(mountWrapper.find('#payment-calculator-modal').length).toBeGreaterThan(0);
  });

  it('posts messages to the server', () => {
    jest.spyOn(axios, 'post');
    mountWrapper.find('button#send-seller-email').simulate('submit');
    expect(axios.post).toHaveBeenCalled();
  });
});
