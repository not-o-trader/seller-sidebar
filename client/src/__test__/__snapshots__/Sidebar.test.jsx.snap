// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> renders 1`] = `
<styled.div>
  <VehiclePrice
    toggleModal={[Function]}
  />
  <hr />
  <SellerInfo />
  <ContactForm
    send={[Function]}
  />
</styled.div>
`;
